`AnswerToSpeech`

ZanSara opened the pull request:
**Proposed changes**:
- Add two new primitives, `AudioAnswer` and `GeneratedAudioAnswer`, that contains references to audio files into their `answer` and `context` fields.
- Add a new node called `AnswerToSpeech` that can convert regular Answers into a `GeneratedAudioAnswer`.

**Open questions**:
- What is your opinion on the new primitives? Is this a direction we would like to take to support multimodal input-output in Haystack?
  - Note, this is a stub: if we like hierarchical primitives, I will do some refinement and come up with a clean hierarchy.
- Shall I add a short tutorial on how to use this node?

**Status (please check what you already did)**:
- [X] First draft (up for discussions & feedback)
- [x] Final code
- [X] [Enable actions on your fork](https://github.com/deepset-ai/haystack/blob/master/CONTRIBUTING.md#forks)
- [x] Added tests
- [X] Updated documentation: https://github.com/deepset-ai/haystack-website/pull/339

julian-risch reviewed (changes requested):
Looks very promising! I left a few comments that we could discuss and I would like to see a small example pipeline/tutorial. Before we can merge this we would need to add docstrings and also test cases so there is still some work to do but so far it looks great! 👍

ZanSara commented on haystack/nodes/audio/answer_to_speech.py:
```diff
@@ -0,0 +1,60 @@
+from haystack.schema import Answer, AudioAnswer, GeneratedAudioAnswer
+
+
+class AnswerToSpeech(BaseComponent):
```
I think that even if we want a node to be able to read documents, we should rather have a auxiliary `TextToSpeech` class that contains the logic, and two `AnswerToSpeech` and `DocumentToSpeech` which can be put in a pipeline. I think a single `TextToSpeech` node would be needlessly complex, given that the input of the two nodes is very different and the node would have to extract the text to convert from different objects. I will prototype the nodes and see what comes out!

ZanSara commented on haystack/nodes/audio/answer_to_speech.py:
```diff
@@ -0,0 +1,60 @@
+import soundfile as sf
+
+from haystack.nodes import BaseComponent
+from haystack.schema import Answer, AudioAnswer, GeneratedAudioAnswer
```
Exactly, I had that in mind. The node is definitely generating "speech", but in the future we can reuse these primitives for other types of audio answers, so we should keep their name generic. So I will keep it like this if you agree this is a reasonable idea

julian-risch replied:
I thought about Speech because right now the AudioAnswer node uses a Text2Speech model under the hood and therefore I found SpeechAnswer more suitable but if we indeed allow Question Answering via extracting a sound snippet from a longer AudioDocument in future then I agree that we can keep AudioAnswer. Similarly, ImageAnswer could be a subimage of a larger Image(Document) then. 
I am trying to imagine how we would generate an AudioAnswer that is not speech based on a TextAnswer. 
Could you maybe come up with an example where we might use an AudioAnswer in future that is not speech?

julian-risch replied:
Oh wow that sounds awesome! I would love to try that out one day. I have got a funny idea what to do with that. :D

ZanSara commented on haystack/schema.py:
```diff
@@ -327,6 +327,42 @@ def from_json(cls, data):
+
+@dataclass
+class GeneratedAudioAnswer(AudioAnswer):
+    type: str = "text-to-speech"
```
Good catch! I will review how the type is used. I originally used `generative` to mimick normal `Answers`, but then I realized this is an even more specific type, so I went for `text-to-speech`.

For now the content of this field is just a label for users to know if the content of the object is extracted or generated, but I don't know how we want to manage them in the future. I can stick to `generative` for now.

julian-risch replied:
What other types could we imagine in future use cases? `extractive` as the `type: Literal["generative", "extractive", "other"]` in `Answer`? Or do you have anything else in mind? If we can come up with some possible values, I would keep the field. Otherwise we might better remove it and then add it again in case a use case comes up in future.

ZanSara commented on haystack/nodes/audio/answer_to_speech.py:
```diff
@@ -0,0 +1,60 @@
+
+    def text_to_speech(self, text: str) -> Any:
+        filename = hashlib.md5(text.encode("utf-8")).hexdigest()
+        path = self.generated_audio_path / f"{filename}.wav"
```
I went for `generated_audio_file`, is it clearer in your opinion?

julian-risch replied:
I would suggest we rename `generated_audio_path` to `generated_audio_dir` and then `path` could become `generated_audio_file_path` or maybe just `file_path`? In our code base `file_path` is used so frequently and never for models so I think `file_path` would be okay - despite my earlier comment.

ZanSara commented on haystack/nodes/audio/answer_to_speech.py:
```diff
@@ -0,0 +1,60 @@
+
+        # Duplicate answers might be in the list, in this case we save time by not regenerating.
+        if not os.path.exists(path):
+            output = self.model(text)["wav"]
```
This is just the model's output: when the file is saved, it should be enough to specify another `audio_format` to obtain compressed audio. I will test this though.

julian-risch commented on haystack/nodes/audio/answer_to_speech.py:
```diff
@@ -0,0 +1,74 @@
+        :param model_name_or_path: the text to speech model, for example `espnet/kan-bayashi_ljspeech_vits`
+        :param generated_audio_dir: folder to save the audio file to
+        :param audio_format: the format to save the audio into (wav, mp3, ...)
+        :param subtype: see soundfile.write()
```
I would prefer if we copy over the information from `soundfile.write()` rather than referring to it. Much easier for users to understand what value to set and how it influences the result.

ZanSara replied:
Unfortunately I just noticed that `soundfile.write()` also has almost no documentation on the matter :see_no_evil: In addition, the subtype seems to be dependent on the uncompressed format selected. I will try to find a better explanation of these strings somewhere and add it here

julian-risch commented on haystack/nodes/audio/document_to_speech.py:
```diff
@@ -0,0 +1,71 @@
+        :param model_name_or_path: the text to speech model, for example `espnet/kan-bayashi_ljspeech_vits`
+        :param generated_audio_dir: folder to save the audio file to
+        :param audio_format: the format to save the audio into (wav, mp3, ...)
+        :param subtype: see soundfile.write()
```
Same as above: I would prefer if we copy over the information from `soundfile.write()` rather than referring to it. Much easier for users to understand what value to set and how it influences the result.

julian-risch reviewed (approved):
LGTM! 👍 This is still a bit experimental I'd say and our documentation could mention that. It will be good for us to see how the community uses this feature. We could even ask for feedback on Haystack slack explicitly. Maybe a community member wants to showcase a use case, who knows? Let's make it as simple as possible for the community to try this out. I left a few comments regarding doc strings and I'll add the needs-documentation to make @agnieszka-m aware of this upcoming new pipeline node.

julian-risch commented:
And for the sake of completeness: there are still a few small mypy issues. Nothing big:
```

haystack/schema.py:272: error: Unsupported left operand type for | ("Dict[str, Any]")
haystack/schema.py:405: error: Unsupported left operand type for | ("Dict[str, Any]")
haystack/nodes/audio/document_to_speech.py:67: error: Need type annotation for "results"
haystack/nodes/audio/answer_to_speech.py:67: error: Dict entry 0 has incompatible type "str": "List[Any]"; expected "str": "AudioAnswer"
haystack/nodes/audio/answer_to_speech.py:70: error: Need type annotation for "results"
```

ZanSara commented on haystack/nodes/audio/_text_to_speech.py:
```diff
@@ -0,0 +1,101 @@
+        path: Path,
+        data: np.array,
+        format: str,
+        sample_rate: int,
```
That value depends on the model's output, so it's hard to guess right. I'd rather let the user take a minute to figure out the correct value.

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Make your QA Pipelines talk!\n",
```
can we change capitalization: Make Your QA Pipelines Talk!

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+    "\n",
+    "Question answering works primarily on text, but Haystack provides some features for audio files that contain speech as well.\n",
+    "\n",
+    "In this tutorial we're going to see how to use `AnswerToSpeech` to convert answers into audio files."
```
... tutorial, we're (missing comma)

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+    "collapsed": false
+   },
+   "source": [
+    "### Prepare environment\n",
```
I think it should be h2
## Prepare your environment

ZanSara replied:
I can change that, but I really copy-pasted it straight from Tut1. Do we want them to be of different sizes?

agnieszka-m replied:
ok, then let's leave it for now, I'll have to revisit it at some point though as it's not great for our SEO

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+   "source": [
+    "### Populate the document store\n",
+    "\n",
+    "First of all, we will populate the document store with a simple indexing pipeline. See Tutorial 1 for more details about these steps."
```
can we add a link to Tutorial 1 here?

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Querying\n",
```
Should this be h2 as well?

agnieszka-m replied:
let's leave as is for now

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+    "Now we will create a pipeline very similar to the basic `ExtractiveQAPipeline` of Tutorial 1,\n",
+    "with the addition of a node that converts our answers into audio files!\n",
+    "\n",
+    "Note that if you're doing document retrieval, Haystack provides and a `DocumentToSpeech` node\n",
```
delete "and" after "Haystack provides"

agnieszka-m commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,360 @@
+   },
+   "outputs": [],
+   "source": [
+    "# You can configure how many candidates the reader and retriever shall return\n",
```
...the Reader and Retriever should return.

ZanSara replied:
This also was copied straight from Tut1. I will fix that too.

julian-risch commented on test/nodes/test_audio.py:
```diff
@@ -0,0 +1,120 @@
+    )
+    assert os.path.exists(audio_file)
+    assert audio_file.suffix == ".mp3"
+    # FIXME find a way to make sure the compressed audio is similar enough to the wav version.
```
That's okay. 👍

julian-risch commented on haystack/nodes/audio/_text_to_speech.py:
```diff
@@ -0,0 +1,140 @@
+        if not os.path.exists(generated_audio_dir):
+            os.mkdir(generated_audio_dir)
+
+        filename = audio_naming_function(text)
```
That's nice! 👍

julian-risch commented on tutorials/Tutorial17_Audio.py:
```diff
@@ -0,0 +1,123 @@
+    # Or use a util to simplify the output
+    # Change `minimum` to `medium` or `all` to raise the level of detail
+    print("\n\nSimplified output:\n")
+    print_answers(prediction, details="minimum")
```
Now that we also have `DocumentToSpeech` I was wondering whether we could exemplify its usage also in the tutorial. For example, at the end of the tutorial show how to convert the document that contains the answer to speech outside of the pipeline, just use the node separately, and play the sound. That would make more users aware of `DocumentToSpeech`. What do you think?

ZanSara replied:
Oh right that's a nice idea. I didn't use it because I didn't want to make another whole pipeline for document search :smile: but this is a good way to use it, even though it won't be in a pipeline. Still better than not using it at all :grin: I'll add!

ZanSara replied:
Or maybe I can add it to the indexing pipeline! :eyes: I'll have a try.

julian-risch reviewed (changes requested):
Great improvements! 👍  I left a few comments. Main request for change is that we can play sounds in jupyter notebooks. I tested that already and one of my comments contains the code.

julian-risch commented:
Ah and a small thing regarding the process: Maybe we could link the Docs Website PR here in the issue description in the line where it says "Updated documentation". That way it's easier for the reviewer to check that it's prepared/merged.

ZanSara commented on tutorials/Tutorial17_Audio.ipynb:
```diff
@@ -0,0 +1,359 @@
+    "retriever = BM25Retriever(document_store=document_store)\n",
+    "reader = FARMReader(model_name_or_path=\"deepset/roberta-base-squad2\", use_gpu=True)\n",
+    "answer2speech = AnswerToSpeech(\n",
+    "    model_name_or_path=\"espnet/kan-bayashi_ljspeech_vits\", generated_audio_dir=Path(__file__).parent / \"audio_answers\"\n",
```
I'll totally fix the `__file__` issue (thanks! I never noticed :sweat_smile:), but I don't really want to point to my branch in the installation script. That would break the tutorial as soon as it's merged to master, which is not ideal :sweat_smile: We should test it with the modification you mention, but not put it in the PR.

ZanSara replied:
I simply used a relative path in this line. Should work, I will test it soon :+1:

julian-risch replied:
Sure, sorry, I didn't mean to put it in the PR, just when you try it out yourself on colab. It's a good check to run it there once before merging.

ZanSara commented:
Ok I'll have to drag this PR along a little longer :sweat_smile: 

I tried to use `DocumentToSpeech` in an indexing pipeline and realized suddenly that the primitives I implemented were just not compatible with indexing, so I reverted part of it back. Now there's no more `AudioAnswer` and `AudioDocument`, but `SpeechAnswer` and `SpeechDocument` which inherit from `Answer` and `Document` directly. They have audio paths as accessory information in additional fields, while by comparison, `AudioAnswer` had the `answer` text field replaced with a path to the audio file.

I will leave a proper implementation of generic primitives to another PR, along the rest of the work to be done for `ImageRetriever`. I will also take care of updating the docs.

ZanSara commented on haystack/nodes/audio/answer_to_speech.py:
```diff
@@ -0,0 +1,88 @@
+        transformers_params: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Convert an input Answer into an audio file containing the answer's answer and context read out loud.
```
Yep that sounds better :smile:

ZanSara commented on haystack/nodes/audio/_text_to_speech.py:
```diff
@@ -0,0 +1,140 @@
+    from espnet2.bin.tts_inference import Text2Speech as _Text2SpeechModel
+except OSError as ose:
+    logging.exception(
+        "sndfile not found. The node will most likely crash. Please install soundfile's dependencies (https://python-soundfile.readthedocs.io/en/latest/)"
```
sndfile is a library, so I believe it should not take an article. I can clarify though, like "`sdnfile binary not found" or by writing the full lib name `libsndfile1` (even though that could be confusing for the rare Windows users)

julian-risch reviewed (approved):
LGTM! 👍
